File: frontend/TaskBoard/Support/OrderKey.swift
import Foundation

/// Fractional ordering keys over base-62, a direct port of the server's
/// algorithm (backend/src/orderkey.js; design §10). Bytewise comparison order is
/// identical in SQLite BINARY collation, JSON, and Swift String < for ASCII.
enum OrderKey {
    static let digits = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")
    private static let minDigit: Character = "0"

    static func isValid(_ key: String) -> Bool {
        guard !key.isEmpty, key.count <= 128, key.last != minDigit else { return false }
        return key.allSatisfy { $0.isASCII && ($0.isNumber || ($0.isLetter && $0.asciiValue != nil)) }
    }

    /// Shortest string strictly between `a` and `b`. `a` may be "" (lower
    /// bound); `b` may be nil (upper bound). NEVER traps: replication can
    /// legitimately produce equal neighbor keys (design §10 — concurrent
    /// same-gap inserts, tie-broken by id), so degenerate bounds fall back to
    /// a key after the lower bound instead of crashing mid-drag.
    static func midpoint(_ a: String, _ b: String?) -> String {
        var lower = a
        while lower.last == minDigit { lower.removeLast() }
        var upper: String? = nil
        if var u = b {
            while u.last == minDigit { u.removeLast() }
            if !u.isEmpty { upper = u }
        }
        if let u = upper, lower >= u {
            // Equal or inverted neighbors: any legal key converges via the
            // (orderKey, id) tie-break — append after the lower bound.
            return core(lower, nil)
        }
        return core(lower, upper)
    }

    private static func core(_ a: String, _ b: String?) -> String {
        let aChars = Array(a)
        if let b {
            let bChars = Array(b)
            var n = 0
            while n < bChars.count && (n < aChars.count ? aChars[n] : minDigit) == bChars[n] { n += 1 }
            if n > 0 {
                return String(bChars.prefix(n)) + core(String(aChars.dropFirst(n)), String(bChars.dropFirst(n)))
            }
        }
        let digitA = aChars.isEmpty ? 0 : (digits.firstIndex(of: aChars[0]) ?? 0)
        let digitB: Int
        if let b, let first = b.first {
            digitB = digits.firstIndex(of: first) ?? digits.count
        } else {
            digitB = digits.count
        }
        if digitB - digitA > 1 {
            let mid = Int((0.5 * Double(digitA + digitB)).rounded())
            return String(digits[mid])
        }
        if let b, b.count > 1 {
            return String(b.first!)
        }
        return String(digits[digitA]) + core(String(aChars.dropFirst()), nil)
    }

    /// A key sorting after the current column maximum ("V" for an empty column).
    static func after(_ maxKey: String?) -> String {
        midpoint(maxKey ?? "", nil)
    }

    /// A key strictly between two neighbors (nil-bounded at either end).
    static func between(_ lower: String?, _ upper: String?) -> String {
        midpoint(lower ?? "", upper)
    }
}

File: backend/src/orderkey.js
// Fractional ordering keys over base-62 (BACKEND_DESIGN.md §10).
// Bytewise comparison order: 0-9 < A-Z < a-z, identical in SQLite BINARY
// collation, JSON strings, and Swift String < for pure ASCII.
export const DIGITS = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'
const MIN_DIGIT = '0'

export const ORDER_KEY_RE = /^[0-9A-Za-z]{1,128}$/

// Generated keys never end in the reserved minimum digit, so a strictly-between
// key always exists (the "K"/"K0" prefix-adjacency edge case).
export function isValidOrderKey(key) {
  return typeof key === 'string' && ORDER_KEY_RE.test(key) && !key.endsWith(MIN_DIGIT)
}

// Shortest string strictly between a and b. a may be '' (lower bound of the
// space); b may be null (upper bound). Inputs must not end in '0'.
export function midpoint(a, b) {
  if (b !== null && a >= b) throw new Error(`midpoint: "${a}" >= "${b}"`)
  if (a.endsWith(MIN_DIGIT) || (b !== null && b.endsWith(MIN_DIGIT))) {
    throw new Error('midpoint: inputs must not end in the minimum digit')
  }
  if (b !== null) {
    // Strip the longest common prefix, padding a with the min digit as we go.
    let n = 0
    while ((a.charAt(n) || MIN_DIGIT) === b.charAt(n)) n++
    if (n > 0) return b.slice(0, n) + midpoint(a.slice(n), b.slice(n))
  }
  const digitA = a ? DIGITS.indexOf(a.charAt(0)) : 0
  const digitB = b !== null ? DIGITS.indexOf(b.charAt(0)) : DIGITS.length
  if (digitB - digitA > 1) {
    return DIGITS.charAt(Math.round(0.5 * (digitA + digitB)))
  }
  // First digits are consecutive.
  if (b !== null && b.length > 1) return b.slice(0, 1)
  return DIGITS.charAt(digitA) + midpoint(a.slice(1), null)
}

// Key sorting after the current maximum in a column ('' / null when empty).
// First key in an empty column is 'V'.
export function keyAfter(maxKey) {
  return midpoint(maxKey || '', null)
}
